pkg/util/eventutil/events.go

```diff
@@ -0,0 +1,90 @@
+/*
+Copyright 2021 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package eventutil
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+)
+
+// EventReporter is custom events reporter type which allows user to limit the events
+type EventReporter struct {
+	mux            sync.Mutex
+	recorder       record.EventRecorder
+	reportedEvents map[string]*eventObject
+
+	// report events x times where x is count
+	count int
+
+	// report events after x minutes
+	eventReportAfterMinutes int
+}
+
+// NewEventReporter returns EventReporter object
+func NewEventReporter(recorder record.EventRecorder, maxCountInGivenTime, reportAfter int) *EventReporter {
+	er := &EventReporter{
+		recorder:                recorder,
+		count:                   maxCountInGivenTime,
+		eventReportAfterMinutes: reportAfter,
+	}
+
+	er.reportedEvents = map[string]*eventObject{}
+
+	return er
+}
+
+// Report records a events if eventReportAfterMinutes has passed or events occurred less than count
+func (rep *EventReporter) Report(instance runtime.Object, instanceName, eventType, eventReason, msg string) {
+	rep.mux.Lock()
+	defer rep.mux.Unlock()
```

Comment:
Rook only reconciles a single cephcluster at a time, so it shouldn't be necessary to lock for event reporting

Reply:
Someone might import this from rook and they might need it. So I am in the opinion to keep it.

Reply:
In general, I would say that support for concurrent access to an API is not necessary if it is not intended by the authors to be explicitly supported. If a new caller decides to call an API concurrently that doesn't support it, the caller can choose to lock the resource. For example, a `map` does not put locks internally, the caller has to lock it when needed. Rook doesn't support a public API at the golang package level, so it seems odd to add locking here where it is not added in other helpers in Rook.